Development/ssl/ssl_utils.cpp

```diff
@@ -0,0 +1,282 @@
+
+            X509_ptr x509(PEM_read_bio_X509_AUX(bio.get(), NULL, NULL, NULL), &X509_free);
+            if (!x509)
+            {
+                throw ssl_exception("failed to load cert: PEM_read_bio_X509_AUX failure: " + last_openssl_error());
+            }
+
+            auto sans = details::subject_alt_names(x509.get());
+
+            auto common_name = details::common_name(x509.get());
+            if (common_name.empty())
+            {
+                throw ssl_exception("missing Common Name");
+            }
+
+            auto issuer_name = details::issuer_name(x509.get());
+            if (issuer_name.empty())
+            {
+                throw ssl_exception("missing Issuer Common Name");
+            }
+
+            // X509_get_notAfter returns the time that the cert expires, in Abstract Syntax Notation
+            // According to the openssl documentation, the returned value is an internal pointer which MUST NOT be freed
+#if (OPENSSL_VERSION_NUMBER >= 0x1010000fL)
+            auto not_before = X509_get0_notBefore(x509.get());
+#else
+            auto not_before = X509_get_notBefore(x509.get());
+#endif
+            if (!not_before)
+            {
+                throw ssl_exception("failed to get notBefore: X509_get0_notBefore failure: " + last_openssl_error());
+            }
+#if (OPENSSL_VERSION_NUMBER >= 0x1010000fL)
+            auto not_after = X509_get0_notAfter(x509.get());
+#else
+            auto not_after = X509_get_notAfter(x509.get());
+#endif
+            if (!not_after)
+            {
+                throw ssl_exception("failed to get notAfter: X509_get0_notAfter failure: " + last_openssl_error());
+            }
+
+            auto not_before_time = details::ASN1_TIME_to_time_t(not_before);
+            auto not_after_time = details::ASN1_TIME_to_time_t(not_after);
+
+            return{ common_name, issuer_name, not_before_time, not_after_time, sans };
+        }
+
+        // split certificate chain to list of certificates
+        std::vector<std::string> split_certificate_chain(const std::string& cert_data)
+        {
+            std::vector<std::string> certs;
+            const std::string begin_delimiter{ "-----BEGIN CERTIFICATE-----" };
+            const std::string end_delimiter{ "-----END CERTIFICATE-----" };
+            size_t start = 0;
+            size_t end = 0;
+            do
+            {
+                start = cert_data.find(begin_delimiter, start);
+                end = cert_data.find(end_delimiter, start);
+
+                if (std::string::npos != start && std::string::npos != end)
+                {
+                    certs.push_back(cert_data.substr(start, end - start + end_delimiter.length()));
+                    start = end + end_delimiter.length();
+                }
+
+            } while (std::string::npos != start && std::string::npos != end);
+
+            return certs;
+        }
+
+        // calculate the number of seconds to expire with the given ratio
```

Comment:
This seems like a slightly odd function to me. Combining `cert_information`, `time` and `difftime` seems natural, but doing the `> 0`, `* ratio` and truncate to `int`, less so.

```suggestion
        // calculate the number of seconds until expiry with the given ratio
```